x-pack/filebeat/module/oracle/database_audit/ingest/pipeline.yml

```diff
@@ -0,0 +1,118 @@
+description: Pipeline for parsing checkpoint firewall logs
+processors:
+- set:
+    field: event.ingested
+    value: '{{_ingest.timestamp}}'
+- grok:
+    field: message
+    patterns:
+    - "%{GREEDYDATA:tmp_timestamp}\\\nLENGTH : '%{GREEDYDATA:LENGTH}'\\\n(?m)%{GREEDYDATA:audit}"
+- kv:
+    field: audit
+    field_split: "\\\n(?=[a-zA-Z])"
+    value_split: ":\\S\\d+\\S(?= ')"
+    trim_value: " '"
+    trim_key: " "
+    prefix: oracle.database_audit.
+- grok:
+    field: log.file.path
+    patterns:
+    - "%{BASE10NUM:process.pid}\\_%{BASE10NUM}\\.aud(\\.log)?$"
+# All field names are uppercase by default, converts them to lowercase
+- script:
+    source: "ctx.oracle.database_audit = ctx.oracle.database_audit.entrySet().stream().collect(Collectors.toMap(entry -> entry.getKey().toLowerCase(), Map.Entry::getValue));"
+    lang: painless
+# Replace all field names that has spaces in them with _
+- script:
+    lang: painless
+    source: "ctx.oracle.database_audit = ctx?.oracle?.database_audit.entrySet().stream().collect(Collectors.toMap(e -> e.getKey().replace(' ', '_'), e -> e.getValue()));"
+- gsub:
+    field: "oracle.database_audit.action"
+    pattern: "\\n"
+    replacement: ""
+- gsub:
+    field: "oracle.database_audit.action"
+    pattern: "\\s{2,}"
+    replacement: " "
+- trim:
+    field: "oracle.database_audit.action_number"
+    ignore_missing: true
+# Removes all null values from ctx.*
+- script:
+    lang: painless
+    if: ctx?.oracle?.database_audit != null
+    source: |
+      void handleMap(Map map) {
```

Comment:
Untested but something like this is a little closer to a classical recursive implementation:
```
void removeNulls(Object obj) {
    if (obj instance of Map) {
	for (def x : obj.entrySet()) {
	    if (x.getValue() == null) {
		obj.remove(x.getKey());
		continue;
	    }
	    removeNulls(x);
	}
    }
    if (obj instance of List) {
	for (def x : obj) {
	    removeNulls(x);
	}
    }
}
removeNulls(ctx);
```

Reply:
After discussing a bit around handling both list and maps with @leehinman we decided to keep it as it is for now, might change in the future.